![Logo](docs/source/images/logo.png)
<!--
```{image} docs/source/images/logo.png
:alt: logo
:class: bg-primary
:width: 500px
:align: center
``` -->

# A fast and accurate ancient DNA damage toolkit for metagenomic data

[![Documentation](https://img.shields.io/badge/documentation-latest-blue.svg)](https://metadmg-dev.github.io/metaDMG-core/)
[![Documentation](https://img.shields.io/badge/dashboard-live-blue.svg)](https://metadmg.herokuapp.com/)
![](https://img.shields.io/pypi/v/metadmg)
![](https://img.shields.io/pypi/pyversions/metaDMG)
<!-- ![](https://img.shields.io/pypi/l/metaDMG) -->
<!-- ![](https://img.shields.io/github/workflow/status/metaDMG-dev/metaDMG-core/CI-CD) -->
<!-- ![](https://img.shields.io/pypi/dm/metaDMG) -->
<!-- ![](https://img.shields.io/github/issues-raw/metaDMG-dev/metaDMG-core) -->
<!-- ![](https://img.shields.io/github/issues-closed-raw/metaDMG-dev/metaDMG-core) -->
<!-- ![](https://img.shields.io/github/languages/code-size/metaDMG-dev/metaDMG-core) -->

`metaDMG` is a novel framework and suite of programs for analysing large-scale genomic data especially in the context of environmental DNA. This includes state-of-the-art statistical methods for computing nucleotide misincorporation and fragmentation patterns of even highly complex samples.

For more information, see the [documentation](https://metadmg-dev.github.io/metaDMG-core).
For a quick preview of the interactivity, see the [dashboard](https://metadmg.herokuapp.com).

## Table of contents
* [Getting started](#getting-started)
* [Citing metaDMG](#citing-metadmg)

## Getting Started

For information about how to get started running `metaDMG`, see the section in the [documentation](https://metadmg-dev.github.io/metaDMG-core/getting-started.html).

## Citing `metaDMG`

Here will be more information once our paper is released.
